### PR TITLE
Improve Event Persistence for IntelligencePlugin by Using UserDefaults

### PR DESCRIPTION
**Pull request by JErazo7**
### **Description:**
This PR refines the event handling mechanism in `IntelligencePlugin` by replacing the in-memory `selectionsBuffer` with `UserDefaults` to persist events even when the app is closed. This update ensures that events triggered by App Intents are reliably stored and sent to the Flutter event channel once the app reopens, preventing any loss of events due to app state limitations.

#### Changes:
- Removed `selectionsBuffer` and replaced it with `UserDefaults` for persistent storage of events.
- Modified `push` method to save events directly to `UserDefaults` for reliable storage.
- Updated `onListen` to retrieve and dispatch stored events from `UserDefaults` to the Flutter event channel upon connection.
- Ensures events triggered by App Intents are delivered reliably without requiring the app to be in the background.

This improvement enhances the reliability of event handling for scenarios where the app is closed, making the plugin more robust in handling background events. 

#### Validation:
1) Run the example app provided in the package repository.
2) Use Siri or the Shortcuts app to trigger the App Intents while the app is fully closed.
3) Open the app and verify that all triggered events are received and processed as expected by checking the Flutter event channel.

### CODE REVIEW

**Comment by wrbl606** [comment]
Thanks for your contribution!

Is the goal of this PR to get rid of the 

```swift
static var openAppWhenRun: Bool = true
```

requirement in each AppIntent?

**Comment by JErazo7** [comment]
> Thanks for your contribution!
> 
> Is the goal of this PR to get rid of the
> 
> ```swift
> static var openAppWhenRun: Bool = true
> ```
> 
> requirement in each AppIntent?

@wrbl606 i'm fixing the following issue you can watch in the next video:

https://github.com/user-attachments/assets/7bd88a30-10ce-4f61-bd1e-73fefc305dd0

As you can see, if the example app is closed (not in the background), the intents do not perform any action. The intents only work when the app is opened or in the background.

**Comment by wrbl606** [comment]
Interesting, the in-memory storage should be sufficient to cold boot the app as it only has to hold onto the pushed entries until the Flutter engine is ready. I'm glad your update worked for you but I don't see a reason why the persistent should change anything in the process, and I do see risks of keeping the events for longer than one app session, e.g. sending stale selections to the app.

Thanks for bringing this up, it might be a bug.

File: ios/Classes/IntelligencePlugin.swift [hunk 1]
```diff
@@ -64,28 +64,35 @@ struct PopulateItem: Decodable {
...
     sink = events
+  
+    // Retrieve persisted events from UserDefaults and send them
+    if let storedSelections = UserDefaults.standard.array(forKey: "storedSelections") as? [String] {
+      for selection in storedSelections {
+        sink?(selection)
+      }
+      // Clear persisted events after sending
+      UserDefaults.standard.removeObject(forKey: "storedSelections")
+    }
```
Comment (wrbl606) [inline ios/Classes/IntelligencePlugin.swift:89]: That's the actual fix of the Flutter engine <-> AppIntent interaction race condition -- pushing the selections stored in `selectionsBuffer` in `onListen` method would do the trick too.

File: ios/Classes/IntelligencePlugin.swift [hunk 2]
```diff
@@ -64,28 +64,35 @@ struct PopulateItem: Decodable {
...
+  
     return nil
   }
   
   public func onCancel(withArguments arguments: Any?) -> FlutterError? {
+    sink = nil
```
Comment (wrbl606) [inline ios/Classes/IntelligencePlugin.swift:95]: Great spot!

**Review by wrbl606 — COMMENTED** [review]
Thank you for your report and the proposed fix. I've identified the root cause of the issue below and will apply the following minimal fix to the plugin's code:

```swift
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    sink = events
    flushSelectionsBuffer(events)
    return nil
  }
```
The issue arose from my incorrect assumption that the Flutter engine would always initialise and connect to the Swift plugins before the .push method is called. In reality, there is no guarantee that the Flutter engine will start first, as it depends on various factors such as the device model or even the method used to trigger the intent (e.g., Siri command, shortcut in a workflow, shortcut added as an action to the home screen, etc.). Additionally, other plugins used by the Flutter app can also influence this behaviour.

I have implemented a fix for the issue and will release it in version 0.1.0. I will not be merging your code at this time, as persistent storage introduces certain risks and, as it turns out, is not necessary for this solution. However, I will acknowledge your contribution in the changelog, as your input was instrumental in identifying and resolving the bug, which could have potentially disrupted many workflows in the future.

**Comment by JErazo7** [comment]
@wrbl606 thanks for working on this and finding the root cause. Feel free to close my PR. I will be waiting for the new version in pub.dev. 🤝

**Comment by wrbl606** [comment]
0.1.0 is now available on pub.dev, thanks again!

**Pull request closed without merging**